pygen/pygen_src/riscv_instr_stream.py

```diff
@@ -0,0 +1,235 @@
+        riscv_instr_stream.__init__(self)
+        self.kernel_mode = 0
+        self.allowed_instr = []
+        self.category_dist = []
+
+    def create_instr_instance(self):
+        for i in range(self.instr_cnt):
+            self.instr_list.append(None)
+
+    def setup_allowed_instr(self, no_branch=0, no_load_store=1):
+        self.allowed_instr = riscv_instr_ins.basic_instr
+        if no_branch == 0:
+            self.allowed_instr.append(
+                riscv_instr_ins.instr_category[riscv_instr_category_t.BRANCH.name])
+        if no_load_store == 0:
+            self.allowed_instr.append(
+                riscv_instr_ins.instr_category[riscv_instr_category_t.LOAD.name])
+            self.allowed_instr.append(
+                riscv_instr_ins.instr_category[riscv_instr_category_t.STORE.name])
+        self.setup_instruction_dist(no_branch, no_load_store)
+
+    # TODO
+    def randomize_avail_regs(self):
+        pass
+
+    def setup_instruction_dist(self, no_branch=0, no_load_store=1):
+        if cfg.dist_control_mode:
+            self.category_dist = cfg.category_dist
+            if no_branch:
+                self.category_dist[riscv_instr_category_t.BRANCH.name] = 0
+            if no_load_store:
+                self.category_dist[riscv_instr_category_t.LOAD.name] = 0
+                self.category_dist[riscv_instr_category_t.STORE.name] = 0
+            logging.info("setup_instruction_dist: %0d", category_dist.size())
+
+    def gen_instr(self, no_branch=0, no_load_store=1, is_debug_program=0):
+        self.setup_allowed_instr(no_branch, no_load_store)
+        for i in range(len(self.instr_list)):
+            self.instr_list[i] = self.randomize_instr(self.instr_list[i], is_debug_program)
+        while self.instr_list[-1].category == riscv_instr_category_t.BRANCH:
+            self.instr_list.pop()
+            if len(self.instr_list):
+                break
+
+    def randomize_instr(self, instr, is_in_debug=0, disable_dist=0):
+        exclude_instr = []
+        is_SP_in_reserved_rd = riscv_reg_t.SP in self.reserved_rd
+        is_SP_in_reserved_regs = riscv_reg_t.SP in cfg.reserved_regs
+        is_SP_in_avail_regs = riscv_reg_t.SP in self.avail_regs
+        if ((is_SP_in_reserved_rd or is_SP_in_reserved_regs) or (not is_SP_in_avail_regs)):
+            exclude_instr.append(riscv_instr_name_t.C_ADDI4SPN)
+            exclude_instr.append(riscv_instr_name_t.C_ADDI16SP)
+            exclude_instr.append(riscv_instr_name_t.C_LWSP)
+            exclude_instr.append(riscv_instr_name_t.C_LDSP)
+        if is_in_debug and (not cfg.enable_ebreak_in_debug_rom):
+            exclude_instr.append(riscv_instr_name_t.EBREAK)
+            exclude_instr.append(riscv_instr_name_t.C_EBREAK)
+        instr = riscv_instr_ins.get_rand_instr(
+            include_instr=self.allowed_instr, exclude_instr=exclude_instr)
+        instr = self.randomize_gpr(instr)
+        return instr
+
+    def randomize_gpr(self, instr):

```

Comment:
please add a TODO here as well, we may want to replace this with the chosen randomization method.